recipes-extended/xen/blktap3/fix-run-time-errors-and-memory-leaks.patch

```diff
@@ -0,0 +1,270 @@
+Index: blktap/tapback/frontend.c
+===================================================================
+--- blktap.orig/tapback/frontend.c
++++ blktap/tapback/frontend.c
+@@ -478,7 +478,7 @@ tapback_backend_handle_otherend_watch(ba
+ {
+     vbd_t *device = NULL;
+     int err = 0, state = 0;
+-    char *s = NULL, *end = NULL, *_path = NULL;
++    char *s = NULL, *end = NULL;
+ 
+ 	ASSERT(backend);
+     ASSERT(path);
+@@ -509,30 +509,7 @@ tapback_backend_handle_otherend_watch(ba
+      */
+ 	s = tapback_xs_read(device->backend->xs, XBT_NULL, "%s",
+ 			device->frontend_state_path);
+-    if (!s) {
+-        err = errno;
+-		/*
+-         * If the front-end XenBus node is missing, the XenBus device has been
+-         * removed: remove the XenBus back-end node.
+-		 */
+-		if (err == ENOENT) {
+-            err = asprintf(&_path, "%s/%s/%d/%d", XENSTORE_BACKEND,
+-                    device->backend->name, device->domid, device->devid);
+-            if (err == -1) {
+-                err = errno;
+-                WARN(device, "failed to asprintf: %s\n", strerror(err));
+-                goto out;
+-            }
+-            err = 0;
+-            if (!xs_rm(device->backend->xs, XBT_NULL, _path)) {
+-                if (errno != ENOENT) {
+-                    err = errno;
+-                    WARN(device, "failed to remove %s: %s\n", path,
+-                            strerror(err));
+-                }
+-            }
+-		}
+-    } else {
++    if(s) {
+         state = strtol(s, &end, 0);
+         if (*end != 0 || end == s) {
+             WARN(device, "invalid XenBus state '%s'\n", s);
+@@ -541,9 +518,7 @@ tapback_backend_handle_otherend_watch(ba
+             err = frontend_changed(device, state);
+     }
+ 
+-out:
+     free(s);
+-    free(_path);
+     return err;
+ }
+ 
+Index: blktap/tapback/tapback.c
+===================================================================
+--- blktap.orig/tapback/tapback.c
++++ blktap/tapback/tapback.c
+@@ -171,6 +171,7 @@ tapback_backend_destroy(backend_t *backe
+         backend->xs = NULL;
+     }
+ 
++    close(backend->ctrl_sock);
+     unlink(backend->local.sun_path);
+ 
+ 	list_del(&backend->entry);
+Index: blktap/vhd/lib/libvhd.c
+===================================================================
+--- blktap.orig/vhd/lib/libvhd.c
++++ blktap/vhd/lib/libvhd.c
+@@ -1333,7 +1333,7 @@ vhd_find_parent(vhd_context_t *ctx, cons
+ 	location   = NULL;
+ 	*_location = NULL;
+ 
+-	if (!parent)
++	if (!parent || !strcmp(parent, ""))
+ 		return -EINVAL;
+ 
+ 	if (parent[0] == '/') {
+@@ -1868,7 +1868,7 @@ vhd_footer_offset_at_eof(vhd_context_t *
+ {
+ 	int err;
+ 	if ((err = vhd_seek(ctx, 0, SEEK_END)))
+-		return errno;
++		return err;
+ 	*off = vhd_position(ctx) - sizeof(vhd_footer_t);
+ 	return 0;
+ }
+@@ -2751,7 +2751,7 @@ vhd_initialize_header_parent_name(vhd_co
+ 	 * MICROSOFT_COMPAT
+ 	 * big endian unicode here 
+ 	 */
+-	cd = iconv_open(UTF_16BE, "ASCII");
++	cd = iconv_open("UTF-16BE", "ASCII");
```

Comment:
This doesn't seem necessary since it's defined in `include/libvhd.h`

Reply:
You're right, updating accordingly.